src/modules/job-manager/jobtap.c

```diff
@@ -0,0 +1,520 @@
+            rc = -1;
+        }
+        if (priority == -1) {
+           /*
+            *  A plugin callback was called but didn't provide a
+            *   priority. This could be due to a loaded plugin that is
+            *   not a priority plugin. Therefore take the default action
+            *   and set priority to job->urgency.
+            */
+            priority = job->urgency;
+        }
+        else if (priority == FLUX_JOBTAP_PRIORITY_UNAVAIL) {
+            /*
+             *  Plugin cannot determine priority at this time. Set
+             *   priority to the current job->priority so that a priority
+             *   event is not generated.
+             */
+            priority = job->priority;
+            /*
+             *  A plugin cannot return an "unavailable" priority from the
+             *   priority.get callback for jobs in SCHED state. Log an error
+             *   in this case and make no change to priority.
+             */
+            if (job->state == FLUX_JOB_STATE_SCHED)
+                flux_log (jobtap->ctx->h, LOG_ERR,
+                          "jobtap: %s: %ju: BUG: plugin didn't return priority",
+                          jobtap_plugin_name (jobtap->plugin),
+                          (uintmax_t) job->id);
+        }
+        /*
+         *   O/w, plugin provided a new priority.
+         */
+    }
+    else if (rc == 0) {
+        /*
+         *  No priority.get callback was run. Enable default behavior
+         *   (priority == urgency)
+         */
+        priority = job->urgency;
+    }
+    else {
+        /*
+         *  priority.get callback was run and failed. Log the error
+         *   and return the current priority.
+         */
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: job.priority.get: callback failed",
+                  jobtap_plugin_name (jobtap->plugin));
+        priority = job->priority;
+    }
+
+    flux_plugin_arg_destroy (args);
+    *pprio = priority;
+    return rc;
+}
+
+int jobtap_call (struct jobtap *jobtap,
+                 struct job *job,
+                 const char *topic,
+                 const char *fmt,
+                 ...)
+{
+    int rc = -1;
+    json_t *note = NULL;
+    flux_plugin_arg_t *args;
+    int64_t priority = -1;
+    va_list ap;
+
+    if (!jobtap->plugin) {
+        /*
+         * Default with no plugin: ensure we advance past PRIORITY state
+         */
+        if (job->state == FLUX_JOB_STATE_PRIORITY)
+            reprioritize_job (jobtap->ctx, job, job->urgency);
+        return 0;
+    }
+
+    va_start (ap, fmt);
+    if (!(args = jobtap_args_vcreate (jobtap, job, fmt, ap))) {
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: %s: %ju: failed to create plugin args",
+                  jobtap_plugin_name (jobtap->plugin),
+                  topic,
+                  (uintmax_t) job->id);
+        return -1;
+    }
+    va_end (ap);
+
```

Comment:
Apparently `va_start()` can allocate memory, so `va_end()` so this function should not return between the two.

https://stackoverflow.com/questions/32259543/is-it-okay-to-longjmp-before-calling-va-end